src/base/hide_password.cpp

```diff
@@ -0,0 +1,861 @@
+		{
+		  assert (0);
+		  return;
+		}
+
+	      memcpy (pwd_info_ptr, hide_pwd_ptr->pwd_info_ptr, hide_pwd_ptr->used * sizeof (int));
+	      if (hide_pwd_ptr->pwd_info_ptr != hide_pwd_ptr->pwd_info)
+		{
+		  free (hide_pwd_ptr->pwd_info_ptr);
+		}
+
+	      hide_pwd_ptr->size = new_size;
+	      hide_pwd_ptr->pwd_info_ptr = pwd_info_ptr;
+	    }
+
+	  assert (SET_PWD_LENGTH (start, end) < MAX_PWD_LENGTH);
+	  hide_pwd_ptr->pwd_info_ptr[hide_pwd_ptr->used++] = start;
+	  hide_pwd_ptr->pwd_info_ptr[hide_pwd_ptr->used++] = SET_PWD_LENGTH_N_ADDINFO (start, end, is_add_comma,
+	      en_add_pwd_string);
+	}
+
+      newptr = skip_one_query (newptr);
+    }
+}
+
+bool
+hide_password::check_capitalized_keyword_create (char *query)
+{
+  char *ps = query;
+  int len;
+  char *token;
+
+  // "create user ~" or " ~; create user ~"
+  token = get_token (ps, len);
+  if ((len == 6) &&  (strncmp (token, "CREATE", len) == 0))
+    {
+      return true;
+    }
+  return false;
+}
+
+const char *
+hide_password::get_password_string (char *qryptr, int *pwd_info_ptr)
+{
+  EN_ADD_PWD_STRING en_pwd_string;
+  // *INDENT-OFF*
+  static const char* password_string[6] = { 
+        ", '****'" ,          " '****'",
+        " PASSWORD '****'" ,  " password '****'",
+        ", PASSWORD='****'" , " password='****'"
+  };
+  // *INDENT-ON*
+
+  en_pwd_string = IS_PWD_NEED_PASSWORD (pwd_info_ptr);
+  if (en_pwd_string == en_none_password)
+    {
+      return (IS_PWD_NEED_COMMA (pwd_info_ptr) ? password_string[0] : password_string[1]);
+    }
+  else
+    {
+      bool is_capatalized = check_capitalized_keyword_create (qryptr);
+
+      if (en_pwd_string == en_user_password)
+	{
+	  return (is_capatalized ? password_string[2] : password_string[3]);
+	}
+      else
+	{
+	  assert (en_pwd_string == en_server_password);
+	  return (is_capatalized ? password_string[4] : password_string[5]);
+	}
+    }
+}
+
+int
+hide_password::snprint_password (char *msg, int size, char *query, HIDE_PWD_INFO_PTR hide_pwd_ptr)
+{
+  char *qryptr = query;
+  char chbk;
+  int pos;
+  int length = 0;
+
+  assert (hide_pwd_ptr);
+  int *pwd_info_ptr = hide_pwd_ptr->pwd_info_ptr;
+
+  for (int x = 0; x < hide_pwd_ptr->used; x += 2)
+    {
+      pos = pwd_info_ptr[x];
+      chbk = query[pos];
+      query[pos] = '\0';
+
+      length += snprintf (msg + length, size - length, "%s", qryptr);
+      length += snprintf (msg + length, size - length, "%s", get_password_string (qryptr, pwd_info_ptr + x));
+
+      query[pos] = chbk;
+      qryptr = query + GET_END_PWD_OFFSET (pwd_info_ptr + x);
+    }
+
+  if (*qryptr)
+    {
+      length += snprintf (msg + length, size - length, "%s", qryptr);
+    }
+
+  return length;
+}
+
+void
+hide_password::fprintf_replace_newline (FILE *fp, char *query, int (*cas_fprintf) (FILE *, const char *, ...))
+{
+  int offset;
+  char chbk;
+
+  while (*query)
+    {
+      offset = strcspn (query, "\r\n");
+      if (offset <= 0)
+	{
+	  cas_fprintf (fp, " ");
```

Comment:
cas_fprintf () in broker/cas_log.c
what about check whether fp is not NULL.
```
if (fp != NULL)
  {
    cas_fprintf (fp, " ");
  }

```

Reply:
Currently, it is guaranteed that fp is not null during the function call.
We will add assert() in preparation for future modifications.